Remove unneeded broadcast() overload.

atgeirr opened the pull request:
Addresses the same problem as #6173.

Note that this changes the order of arguments for some calls: now the root always comes first, there is no possible mixup between the root argument and the data arguments.

Most changes are due to no longer having a default argument for the root. I consider that a good thing in itself.

There was only a single case where the two-argument version of the deleted overload was used, so I actually had to change the order, and that was in MPISerializer itself.

atgeirr commented:
jenkins build this serial please

blattms commented:
mmh, now that I read it, it seem to looks strange somebody that used MPI_Broadcast before There root an argument after the data to send.

Hence at least for broadcasting PODs developers might be more tempted to use the wrong order now.

atgeirr commented:
> Hence at least for broadcasting PODs developers might be more tempted to use the wrong order now.

That is not new here. The variadic overload has that order and cannot be changed (and kept variadic). And the ambiguity must be resolved anyway.

I think the remaining solution here is to add a typing to this:
```c++
struct RootRank
{
    int value;
};
```
Then we make broadcast() take a `RootRank` instead of an `int`, and call with
```c++
ser.broadcast(RootRank{0}, other, arguments);
```

Would this be a good enough solution?

akva2 commented:
while i am usually not too fond of default arguments, here it makes a lot of sense to me as in 99.9% of cases this will have the defaulted value, using anything else is kinda fringe. this is why I did the overload in the first place. But if it is an issue for clang to understand what I consider perfectly valid code, (ie, empty variadic is the last case it should consider, any other overload should take precedence, something clang and gcc has been able to do for years until recent bump), we have to do something, and I guess this is as good as it gets.

akva2 commented on opm/simulators/wells/WellConnectionAuxiliaryModule.hpp:
```diff
@@ -72,7 +72,7 @@ class WellConnectionAuxiliaryModule : public BaseAuxiliaryModule<TypeTag>
         // Communicate Map to other processes, since it is only available on rank 0
         Parallel::MpiSerializer ser(lin_comm_);
-        ser.broadcast(0, possibleFutureConnections);
+        ser.broadcast(Opm::Parallel::RootRank{0}, possibleFutureConnections);
```
in the Opm namespace.

akva2 commented on opm/simulators/wells/BlackoilWellModelGeneric.cpp:
```diff
@@ -1903,7 +1903,7 @@ getMaxWellConnections() const
     // Communicate Map to other processes, since it is only available on rank 0
     Parallel::MpiSerializer ser(comm_);
-    ser.broadcast(0, possibleFutureConnections);
+    ser.broadcast(Opm::Parallel::RootRank{0}, possibleFutureConnections);
```
In Opm namespace

atgeirr commented:
Updated to use explicit type for the RootRank argument to avoid any chance of mixup.

atgeirr commented:
jenkins build this please

blattms commented:
We currently experiencing strange bugs (when using parameter packs) in our code that are related to the code that is changed here. Please hold off merging until we found out whether our fix also works with this PR

atgeirr commented:
> We currently experiencing strange bugs (when using parameter packs) in our code that are related to the code that is changed here. Please hold off merging until we found out whether our fix also works with this PR

Noted.

In the meantime: @akva2 I addressed your comments except one, about `::Opm::Parallel::`, where I changed both to just `Parallel::`.

atgeirr commented:
jenkins build this serial please

lisajulia commented:
> We currently experiencing strange bugs (when using parameter packs) in our code that are related to the code that is changed here. Please hold off merging until we found out whether our fix also works with this PR

This should can be merged from my side :) Thanks!
